### PR TITLE
Implement embedding ingestion logic

### PR DESCRIPTION
**Pull request by shixiao-coder**
This PR add the top level var spanner_enable_embedding_ingestion that are passed into ingestion_helper as enable_embedding_ingestion and set ingestion_helper environment variable enable_embeddings

It also the logic to do run_postprocessings and run_embeddings in parallel if both are enabled

The last change is that this PR modifies the timeout to be 30 minutes instead of 5 minutes (embedding takes 7 minutes for UN data)

The change is deploy onto a testing [service](https://pantheon.corp.google.com/run/detail/us-central1/xiao2-dc-datacommons-service/observability/metrics?e=13803378&mods=-monitoring_api_staging&project=datcom-website-dev) that works properly

The ingestion [preprocess](https://pantheon.corp.google.com/run/jobs/details/us-central1/xiao2-dc-ingestion-preprocessing-job/executions?project=datcom-website-dev&e=13803378&mods=-monitoring_api_staging) and [processing](https://pantheon.corp.google.com/workflows/workflow/us-central1/xiao2-dc-ingestion-workflow) workflow also finishes properly under different conditions to test post processing both embeddings & aggregation, one of them and none of it

### CODE REVIEW

File: infra/dcp/modules/stack/variables.tf [hunk 1]
```diff
@@ -27,6 +27,7 @@ variable "spanner_config" {
     version_retention_period           = string
     processing_units                   = number
     enable_bigquery_connection         = bool
+    enable_embedding_ingestion         = bool
```
Comment (gmechali) [inline infra/dcp/modules/stack/variables.tf]: I might suggest just calling this enable_embeddings. It's within the spanner_config. Im not sure the "ingestion" makes sense, maybe enable_embeddings_[creation|generation]

File: infra/dcp/variables.tf [hunk 2]
```diff
@@ -191,6 +191,12 @@ variable "spanner_enable_bigquery_connection" {
   default     = true
 }
 
+variable "spanner_enable_embedding_ingestion" {
+  description = "Enable embedding ingestion in Spanner"
```
Comment (gmechali) [inline infra/dcp/variables.tf]: nit: Enable embeddings generation in Spanner.

File: infra/dcp/modules/ingestion/workflow/main.tf [hunk 3]
```diff
@@ -110,10 +110,40 @@ resource "google_workflows_workflow" "ingestion_orchestrator" {
               - check_success:
                   switch:
                     - condition: '$${job_status.currentState == "JOB_STATE_DONE"}'
-                      next: %{if var.enable_bigquery_postprocessing}run_postprocessings%{else}promote_version%{endif}
+                      next: %{ if var.enable_bigquery_postprocessing && var.enable_embedding_ingestion }run_optional_tasks%{ else }%{ if var.enable_bigquery_postprocessing }run_postprocessings%{ else }%{ if var.enable_embedding_ingestion }run_embeddings%{ else }promote_version%{ endif }%{ endif }%{ endif }
               - fail_on_job_status:
                   raise: '$${ "Dataflow job failed with state: " + job_status.currentState }'
```
Comment (gmechali) [inline infra/dcp/modules/ingestion/workflow/main.tf:120]: One more thing, and this could be done in a follow up, we should also make this variable be the controller for disabling the GCS embeddings.

GCS embeddings are generated in here: https://github.com/datacommonsorg/import/blob/89d000b72cfc518640b0cb1a4d7b8b3dac221168/simple/stats/runner.py#L592

In a follow up PR, can you pass in an env var in that job to disable the generate_nl_artifacts when you create spanner embeddings?

File: infra/dcp/modules/ingestion/workflow/main.tf [hunk 4]
```diff
@@ -110,10 +110,40 @@ resource "google_workflows_workflow" "ingestion_orchestrator" {
               - check_success:
                   switch:
                     - condition: '$${job_status.currentState == "JOB_STATE_DONE"}'
-                      next: %{if var.enable_bigquery_postprocessing}run_postprocessings%{else}promote_version%{endif}
+                      next: %{ if var.enable_bigquery_postprocessing && var.enable_embedding_ingestion }run_optional_tasks%{ else }%{ if var.enable_bigquery_postprocessing }run_postprocessings%{ else }%{ if var.enable_embedding_ingestion }run_embeddings%{ else }promote_version%{ endif }%{ endif }%{ endif }
```
[the remark this answers is not in the document]
Comment (gmechali) [reply infra/dcp/modules/ingestion/workflow/main.tf]: Definitely this pls!

File: infra/dcp/modules/ingestion/workflow/main.tf [hunk 5]
```diff
@@ -124,7 +154,21 @@ resource "google_workflows_workflow" "ingestion_orchestrator" {
                       actionType: "run_aggregation"
                       importList: '$${json.decode(input.importList)}'
                   result: postprocessing_result
-%{endif}
+%{ endif }
+%{ if var.enable_embedding_ingestion }
```
Comment (gmechali) [inline infra/dcp/modules/ingestion/workflow/main.tf]: Do you till need this? and lines 146 to 157?

It looks that is duplicated now.

File: infra/dcp/modules/ingestion/helper_service/main.tf [hunk 6]
```diff
@@ -64,6 +64,10 @@ resource "google_cloud_run_v2_service" "ingestion_helper" {
         name  = "IS_BASE_DC"
         value = "false"
       }
+      env {
+        name  = "ENABLE_EMBEDDINGS"
+        value = var.enable_embedding_ingestion
+      }
```
[the remark this answers is not in the document]
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/helper_service/main.tf:70]: the API access is setup at top level

File: infra/dcp/modules/ingestion/workflow/main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/main.tf]: updated with a new name postprocessing_step, it will redirect to a new step name ingestion_postprocessing that the definition is condition based

File: infra/dcp/modules/ingestion/workflow/variables.tf [hunk 7]
```diff
@@ -36,6 +36,16 @@ variable "enable_bigquery_postprocessing" {
...
 
+variable "enable_embedding_ingestion" {
+  type    = bool
+  default = false
+}
+
+variable "embeddings_timeout" {
+  type    = number
+  default = 1800
+}
```
[the remark this answers is not in the document]
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/variables.tf]: added

File: infra/dcp/modules/stack/variables.tf (hunk 1 above)
Comment (shixiao-coder) [reply infra/dcp/modules/stack/variables.tf]: sure, updated to enable_embeddings for all

File: infra/dcp/variables.tf (hunk 2 above)
Comment (shixiao-coder) [reply infra/dcp/variables.tf]: updated

File: infra/dcp/modules/ingestion/workflow/main.tf (hunk 3 above)
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/main.tf:120]: Sure will do this in a follow up, I'd prefer to have the reading part working first before deleting this GCS embeddings

File: infra/dcp/modules/ingestion/workflow/main.tf (hunk 5 above)
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/main.tf]: Just updated the logic to merge into a single step

File: infra/dcp/modules/ingestion/workflow/main.tf [hunk 8]
```diff
@@ -110,21 +115,53 @@ resource "google_workflows_workflow" "ingestion_orchestrator" {
               - check_success:
                   switch:
                     - condition: '$${job_status.currentState == "JOB_STATE_DONE"}'
-                      next: %{if var.enable_bigquery_postprocessing}run_postprocessings%{else}promote_version%{endif}
+                      next: ${local.postprocessing_step}
```
Comment (gmechali) [inline infra/dcp/modules/ingestion/workflow/main.tf]: I think what was here before is a bit clearer than now.

Can you make the variable be
locals.should_run_postprocessing = var.enable_bigquery_postprocessing || var.enable_embeddings_generation;

And here you can just do
 %{if local.should_run_postprocessing}ingestion_postprocessing%{else}promote_version%{endif}

File: infra/dcp/modules/ingestion/workflow/main.tf [hunk 9]
```diff
@@ -110,21 +115,53 @@ resource "google_workflows_workflow" "ingestion_orchestrator" {
...
+                                  url: '${var.ingestion_helper_url}'
+                                  auth:
+                                    type: OIDC
+                                  body:
+                                    actionType: "run_aggregation"
+                                    importList: '$${json.decode(input.importList)}'
+                                result: postprocessing_result
+                            - end_postproc_branch:
+                                assign:
+                                  - dummy: true
```
Comment (gmechali) [inline infra/dcp/modules/ingestion/workflow/main.tf:143]: what is this?

Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/main.tf:143]: we need to have this for the condition that we are not running postprocessing, if we directly jump to promote_version, we are risking doing two promote_version in parallel

File: infra/dcp/modules/ingestion/workflow/main.tf (hunk 8 above)
Comment (shixiao-coder) [reply infra/dcp/modules/ingestion/workflow/main.tf]: updated

**Review by gmechali — APPROVED** [review]
Thank you Xiao!
